irohad/ametsuchi/impl/postgres_specific_query_executor.cpp

```diff
@@ -197,27 +197,49 @@ namespace iroha {
           qry.get());
     }
 
-    template <typename RangeGen, typename Pred>
-    std::vector<std::unique_ptr<shared_model::interface::Transaction>>
+    template <typename RangeGen, typename Pred, typename OutputIterator>
+    iroha::expected::Result<void, std::string>
     PostgresSpecificQueryExecutor::getTransactionsFromBlock(
-        uint64_t block_id, RangeGen &&range_gen, Pred &&pred) {
-      std::vector<std::unique_ptr<shared_model::interface::Transaction>> result;
-      auto block = block_store_.fetch(block_id);
-      if (not block) {
+        uint64_t block_id,
+        RangeGen &&range_gen,
+        Pred &&pred,
+        OutputIterator dest_it) {
+      auto opt_block = block_store_.fetch(block_id);
+      if (not opt_block) {
         log_->error("Failed to retrieve block with id {}", block_id);
-        return result;
+        return {};
       }
+      auto &block = opt_block.value();
+
+      boost::optional<std::string> error;
+      const auto check_tx_index = [this,
+                                   block_id,
+                                   block_size = block->transactions().size(),
+                                   &error](auto tx_idx) {
+        if (tx_idx >= block_size) {
+          error = (boost::format("Failed to retrieve transaction with id %1% "
```

Comment:
Maybe it is better to rewrite the `transform-filtered-transformed` sequence as a simple for-loop? Because this approach looks quite complex.